kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/framework/concurrency/eventually.kt

```diff
@@ -0,0 +1,138 @@
+package io.kotest.framework.concurrency
+
+import io.kotest.assertions.ErrorCollectionMode
+import io.kotest.assertions.errorCollector
+import io.kotest.assertions.failure
+import io.kotest.common.ExperimentalKotest
+import io.kotest.mpp.timeInMillis
+import kotlinx.coroutines.delay
+import kotlin.reflect.KClass
+
+@ExperimentalKotest
+data class EventuallyConfig(
+   val duration: Millis = defaultDuration,
+   val interval: Interval = 25L.fixed(),
+   val retries: Int = Int.MAX_VALUE,
+   val exceptions: Set<KClass<out Throwable>> = setOf(),
+) {
+   init {
+      require(retries > 0) { "Retries should not be less than one" }
+      require(duration > 0L) { "Duration cannot be negative" }
+   }
+
+   companion object {
+      const val defaultDuration: Millis = 3_600_000L
+   }
+}
+
+@ExperimentalKotest
+data class EventuallyState<T>(
+   val result: T?,
+   val start: Instant,
+   val end: Instant,
+   val iteration: Int,
+   val firstError: Throwable?,
+   val thisError: Throwable?,
+)
+
+@ExperimentalKotest
+fun interface EventuallyListener<T> {
+   fun onEval(state: EventuallyState<T>): Boolean
+
+   companion object {
+      val default = EventuallyListener<Any?> { it.thisError == null }
+   }
+}
+
+@ExperimentalKotest
+suspend fun <T> eventually(
+   duration: Millis,
+   interval: Interval = 25L.fixed(),
+   retries: Int = Int.MAX_VALUE,
+   exceptions: Set<KClass<out Throwable>> = setOf(),
+   listener: EventuallyListener<T>? = null,
+   f: ConcurrencyProducer<T>
```

Comment:
We could drop eventually supporting return values entirely, and roll them into until ?

until supports booleans, but I wonder if until could be changed to use T rather than Boolean and it just work as is for booleans?

Then it might all be a bit neater ?

Reply:
I'll have to think about this, we'd have to move the exception set to until as well since right now it doesn't try to catch anything iirc